### ISSUE TITLE
Cython class with stub reported as disjoint base with mypy 1.18.x

### ISSUE THREAD

**Issue by djhoese** [python/mypy#20115]
**Bug Report**

The pykdtree package has a Cython module with a single class definition. Recently a stub was added for this class in a separate `kdtree.pyi` file (see below). Starting with mypy 1.18.x the stubtest has started failing because mypy considers it a disjoint base, but I'm not sure how that is possible since it is a single class. I have very little knowledge of how mypy investigates a C-level class like this so we could easily be doing something wrong.

**To Reproduce**

I have a Cython class that you can find the full source for at:

https://github.com/storpipfugl/pykdtree/blob/master/pykdtree/kdtree.pyx

And stub file:

https://github.com/storpipfugl/pykdtree/blob/master/pykdtree/kdtree.pyi

Here's the init-level part of the Cython class:

```cython
cdef class KDTree:

    def __cinit__(KDTree self):
        ...

    def __init__(KDTree self, np.ndarray data_pts not None, int leafsize=16):
        ...
```

And the relevant portion of the stub file:

```python
class KDTree:

    def __init__(self, data_pts: np.ndarray, leafsize: int = 16): ...
```

Since `__cinit__` is a C-level only method it is not in the stub file.

**Expected Behavior**

Testing the stub with "stubtest" should report no issues. This is true for mypy 1.17.x.

**Actual Behavior**

```
error: pykdtree.kdtree.KDTree is a disjoint base at runtime, but isn't marked with @disjoint_base in the stub
Stub: in file /home/davidh/repos/git/pykdtree/pykdtree/kdtree.pyi:21
<TypeInfo pykdtree.kdtree.KDTree>
Runtime:
<class 'pykdtree.kdtree.KDTree'>
```

Note: I did not write the stub file for this, but am one of the maintainers of pykdtree. I don't fully understand the best way to stub a Cython file, but it used to work so I'm surprised by this.

Possibly Related: https://github.com/python/mypy/issues/19877 https://github.com/python/mypy/issues/19887 https://github.com/python/mypy/pull/19740

**Your Environment**

- Mypy version used: 1.18.2 (and 1.18.1)
- Mypy command-line flags: `stubtest pykdtree`
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.12

**Comment by bzoracler** [comment]
This is the relevant part of `stubtest` that checks for disjoint bases:

https://github.com/python/mypy/blob/fb16e938d12f6b24b5edf8023d4adf8e0e36abb4/mypy/stubtest.py#L481-L511

From a glance at the Cython code (specifically [these lines](https://github.com/storpipfugl/pykdtree/blob/2009fd6e3e8c6a9850c71f41484f04d1cb4303c5/pykdtree/kdtree.pyx#L106-L129)), the `cdef readonly` members probably cause stubtest to flag `KDTree` as a disjoint base. (`readonly` is exposed to Python as readable attributes, so has the same effect as filling these members in `__slots__` in terms of disjoint base detection.)

**Comment by JelleZijlstra** [comment]
See https://peps.python.org/pep-0800/ for background on disjoint bases. I haven't tried but I suspect the error is correct and this cdef class is indeed a disjoint base. You can confirm by checking whether something like `class X(int, KDTree): pass` succeeds in creating a class; if it doesn't, KDTree is likely a disjoint base.

**Comment by djhoese** [comment]
```
In [3]: class MyKDTree(int, KDTree):
   ...:     pass
   ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 class MyKDTree(int, KDTree):
      2     pass

TypeError: multiple bases have instance lay-out conflict

```

(Python 3.12)

Wow I really don't understand this, but what I'm hearing is "readonly cython properties make this disjoint" then 👍 

I understand the idea that some classes can not be inherited from in the same sub-class, but defining that from the parent class definition confuses me. For example, why can't I define a subclass of KDTree that also inherits from a "mix-in" class that just adds another method? I'm not sure my brain on a Friday afternoon is ready to learn.

**Comment by bzoracler** [comment]
Whoops, I think I'm mistaken - it appears that all `cdef` variables in a Cython class (not just `readonly` ones) count towards disjoint base checks.

But in any case, a disjoint-based class tells you what a class can't be subclassed with (they can't be subclassed with other disjoint based classes). You *can* define a subclass of `KDTree` that also inherits from a mix-in that just adds another method. You just need to make sure that the mix-in isn't a disjoint base (e.g., it doesn't define non-empty `__slots__`, and it isn't another Cython class which has extra members).

So, this will work:

```python
class Mixin:
    def another_method(self) -> None: ...

class MyKDTree(KDTree, Mixin): pass
```

But this doesn't:

```python
class MixinWithAnotherAttribute:
    __slots__ = ("another_attribute",)

    def another_method(self) -> None: ...

class MyKDTree(KDTree, MixinWithAnotherAttribute): pass
```

Back to your original issue, you just need to add this on `KDTree` in the stub:

```python
from typing_extensions import disjoint_base

@disjoint_base
class KDTree: ...
```

**Comment by djhoese** [comment]
Ok I think that makes sense. Thanks. I guess this can be closed? We've added that decorate to pykdtree now.
